_posts/2019-11-19-demo-fraud-detection.md

```diff
@@ -0,0 +1,223 @@
+                        Details will be discussed in a future blog post. */;
+
+  @Override
+  public void processElement(
+      Transaction event,
+      Context ctx,
+      Collector<Keyed<Transaction, String, Integer>> out) {
+
+      for (Rule rule :rules) {
+       out.collect(
+           new Keyed<>(
+               event,
+               KeysExtractor.getKey(rule.getGroupingKeyNames(), event),
+               rule.getRuleId()));
+      }
+  }
+  ...
+}
+```
+ `KeysExtractor.getKey()` uses reflection to extract the required values of `groupingKeyNames` fields from events and combines them as a single concatenated String key, e.g `"{beneficiaryId=25;payeeId=12}"`. Flink will calculate a hash of this key and effectively pin processing of this particular combination to a specific server in the cluster. This will allow tracking all transactions between _beneficiary #25_ and _payee #12_ and evaluating defined rules within a desired time window.
+
+Notice that a wrapper class `Keyed` with the following signature was introduced as the output type of `DynamicKeyFunction`:  
+
+```java   
+public class Keyed<IN, KEY, ID> {
+  private IN wrapped;
+  private KEY key;
+  private ID id;
+
+  ...
+  public KEY getKey(){
+      return key;
+  }
+}
+```
+
+Fields of this POJO carry the following information: `wrapped` is the original transaction's event data, `key` is the result of using `KeysExtractor` and `id` is the ID of the Rule which caused the dispatch of the event according to this Rule's grouping logic.
+
+Events of this type will be the input to the `keyBy()` function in the main processing pipeline. This allows us to use a simple lambda-expression as a [`KeySelector`](https://ci.apache.org/projects/flink/flink-docs-stable/dev/api_concepts.html#define-keys-using-key-selector-functions) for the final step of implementing dynamic data shuffle.
+
+```java
+DataStream<Alert> alerts =
+    transactions
+        .process(new DynamicKeyFunction())
+        .keyBy((keyed) -> keyed.getKey());
+        .process(new DynamicAlertFunction())
+
+        //.process(/* actual calculations and alerting */)
+```
+
+By applying `DynamicKeyFunction` we are effectively copying events for performing parallel per-rule evaluation within a Flink cluster. By doing so, we achieve an important property - horizontal scalability of rules' processing. Our system will be capable of handling more rules by adding more servers to the cluster, i.e. increasing the parallelism. This property is achieved at the cost of data duplication, which might become an issue depending on the specific set of parameters, such as incoming data rate, available network bandwidth, event payload size etc. In a real-life scenario, additional optimizations can be applied, such as combined evaluation of rules which have the same `groupingKeyNames`, or a filtering layer, which would strip events of all the fields that are not required for processing of a particular rule.
```

Comment:
```suggestion
With the application of `DynamicKeyFunction` we effectively copy events for performing parallel per-rule evaluation within a Flink cluster. By doing so, we achieve an important property: horizontal scalability of rule processing. Our system will be capable of handling more rules by adding more servers to the cluster, i.e. increasing the parallelism. This property is achieved at the cost of data duplication, something that might become an issue depending on the specific set of parameters, such as incoming data rate, available network bandwidth, event payload size, etc. In a real-life scenario, additional optimizations can be applied, such as a combined evaluation of rules which have the same `groupingKeyNames`, or a filtering layer, which would strip events from all the non-required fields for the processing of a particular rule.
```

Reply:
I believe "strip of" is correct.

Reply:
I think you mean strip off? 

I updated the suggestion accordingly. 
Let me know if that's better now!

Reply:
@sjwiesman ? :)